### ISSUE TITLE
Problem for Far Manager v3.0 b2927+ - failse detection function as variable on Git- 78fc1cf

### ISSUE THREAD

**Issue by VictorVG** [mpeterv/luacheck#97]
Far Manager 3.0 b2927 or letter used Lua and Moonscript as program languages for write scripts and plugins. See as samples my simple test script:
```
far.Message("Begin FarColorer status test");
local CLRG = "D2F36B62-A470-418d-83A3-ED7A3710E5B5";
local CLRS = Plugin.Call(CLRG,"status");

far.Show("Status before all tests",Plugin.Call(CLRG,"status"))

  Plugin.Call(CLRG,"status",0)
   far.Show("Test command Disable",Plugin.Call(CLRG,"status"))

  Plugin.Call(CLRG,"status",1)
   far.Show("Test command Enable",Plugin.Call(CLRG,"status"))

  Plugin.Call(CLRG,"status",0)
   Plugin.Call(CLRG,"status",CLRS)
   far.Show("Test restore status",Plugin.Call(CLRG,"status"))

far.Message("All tests complted");
   far.Show("Current status after all tests",Plugin.Call(CLRG,"status"))
```
this script used internal Far function far.*() and standard Far function for manage plugins Plugin.*() ( please, see http://svn.code.sf.net/p/farmanager/code/trunk/enc/ for details, I'm sorry, but I have no right without the permission of Far Group give you compiled in CHM documentation file) and LuaCheck v0.18 on this script can't find any warning, but v0.19 DEV have false detect for all function like far.Show(), Plugn.Call(), mf.Menu() - and it looks like it does not check for records of the operator that is a function rather than a variable - just do not know, then we consider all the variables and print false warning. :)

Personally, I think that in the case of operators similar to writing on the function of checking algorithm should be considered a priority of their correct function declaration, and only if, during the analysis of the code by the preprocessor turns out that this is a declaration of the array to check on the correctness of its elements - the visibility is correct declaration and other ...

For Your test's I make minimal test kit - x86 for Windows XP SP3 and x64 for Vista or letter (Vista required FarColorer and NetBox x64 plugins)  [**Far3b4904-x86_x64-bin.7z**](https://www.upload.ee/files/6730768/Far3b4904-x86_x64-bin.7z.html) (~10 Mb).

**Comment by VictorVG** [comment]
По русски (наш общий друг подсказал, я не знал, прошу прощения) тут мы с вами встречаемся с ошибочным опознанием вызовов функций в качестве переменных. И тут я бы поступил так, как в своё время это было реализовано в компиляторе IBM PL/1F - если компилятор встречает незнакомую синтаксическую конструкцию, то алгоритм считает её заведомо корректной функцией другого языка и пропускает, если внешне она похожа на вызов функции, а в PL/1 прототип  описывается конструкцией вида:

proc options(func);
....
return
end func;

и вызов call(func);

и почему бы и нам не поступать аналогично? Пока нижележащий код явно не похож на массив считать вызов вида far*(), Plugin.*(), mf.*() корректно определённой функцией другого языка и пропускать её не вывода ошибочный варнинг W113 на каждую незнакомую функцию с составным именем как на примере:

![b4904](https://cloud.githubusercontent.com/assets/906886/23466156/5a89fd04-feab-11e6-91d2-2e2c2700a654.png)

**Comment by mpeterv** [comment]
When Luacheck encounters a global, it looks it up in the list of defined globals. If it is not there, a warning is emitted. Omitting such warnings, regardless of the way the global is used, seems like a bad idea as it would skip warnings for typos such as `ttable.insert()`.

The change that triggered the warnings is that the default list of defined globals is constructed in a different way now. Previously it contained all globals present when Luacheck started up. Now the list contains standard globals of Lua 5.1/5.2/5.3 or LuaJIT based on version of Lua used to run Luacheck.

As a solution, you can define the globals provided by Far Manager using inline options or alter the wrapper used to call Luacheck to define them in the options.

---

Когда Luacheck встречает глобальную переменную, он ищет её в списке разрешённых переменных и выдаёт предупреждение, если её в нём нет. Пропускать такие предупреждения в случае, когда переменная используется каким-то особым способом, это не очень хорошая идея так как тогда не будут обнаруживаться опечатки типа `ttable.insert()`.

Предупреждения стали появляться так как изменился список разрешённых глобальных переменных, используемый по умолчанию. Раньше в него входили все глобальные переменные, установленные на момент старта Luacheck. Теперь в список входят только переменные стандартной библиотеки Lua 5.1/5.2/5.3 или LuaJIT в зависимости от того, в какой версии Lua запускается Luacheck.

Чтобы устранить эти предупреждения, нужно добавить глобальные переменные Far Manager в список разрешённых. Можно использовать настройки в комментариях прямо в коде, или изменить плагин, запускающий Luacheck, чтобы он передавал нужный параметр при запуске.

**Comment by VictorVG** [comment]
Согласен, уже думал что стоит так сделать, только как именно это реализовать идей нет.:( А как вы посоветуете это сделать чтобы корректно добавить такой список?

**Comment by mpeterv** [comment]
Часть Editor_LuaCheck.lua:

```lua
-- luacheck options
local options = { --http://luacheck.readthedocs.org/en/0.11.0/config.html
  new_read_globals={
    "Macro","NoMacro","Event","NoEvent","MenuItem","NoMenuItem","CommandLine","NoCommandLine",
    "IsLuaStateRecreated", --lua plugins
    "_filename",           --scripts executed in luamacro via cmdline prefix

  };
  new_globals={
    "RecreateLuaState",    --lua plugins
  }

  --global=false;
  --allow_defined=true;
  --allow_defined_top=true;
  --unused=false;
  --unused_args=false;
  --unused_secondaries=false;
  --redefined=false;
  --http://luacheck.readthedocs.org/en/0.11.0/cli.html#patterns
  --ignore={"4../.*"};
  --enable={};
  --only={};
}
```

Можно просто добавить имена переменных в массивы `new_read_globals` или `new_globals`, в зависимости от того, разрешается ли изменение их значений. В 0.19.0 можно также полностью описать все разрешённые поля, но пока эта версия не выпущена.

**Comment by VictorVG** [comment]
О! Спасибо огромное! Пока вот смотрю что Кристоф в SumatraPDF поправил - вроде с трудом, но мы смогли его убедить в том, что на Win7 вылезли сразу три ошибки - программа падает при восстановлении сессии, в заголовков табов на ней же не выводится текст с именем файла или документа, ну и третья это просто мелочь - в инсталлере пропала графическая заставка (я ещё три дня назад про это ему писал в баге #729) набор багов #730 (это переоткрытый баг 729 - по словам Кристофа он не смог его воспроизвести), #732, #733. Я лично думаю что причина в "детских болезнях" перехода с С++9 на С++11, ну потратим силы, но отловим и если сможем Кристофу объяснить что происходит он их легче исправит. Ну, или снова закроет типа не может проверить и придётся опять его убеждать в очевидном.:)

**Comment by shmuz** [comment]
> Пока вот смотрю что Кристоф в SumatraPDF поправил

И причём здесь это?

**Comment by VictorVG** [comment]
Просто М$ всем нам там здорово потрепала нервы убрав из списка API в т.ч.  и ГДИ+ - всё упералось в него. Кристоф буквально пару часов назад смог поймать условия когда всё вылезло и поправил. А я уже устал провоцировать дебажную сборку упасть чтобы отладчик сказал где ошибка. Неделя возни и казалось впустую. Ворчу с усталости.

**Comment by shmuz** [comment]
@VictorVG, это всё здесь абсолютный офтопик. Предлагаю вам удалить офтопик, и я в свою очередь удалю свои замечания.

**Comment by VictorVG** [comment]
Спасибо за подсказку! Сделал там где вы предложили:

> -- luacheck options
> local options = { --http://luacheck.readthedocs.io/en/stable/config.html
>   new_read_globals={
>     "Macro","NoMacro","Event","NoEvent","MenuItem","NoMenuItem","CommandLine","NoCommandLine","Plugin","Menu","Dlg","Editor","Panel","Drv","Object","Mouse","CmdLine","APanel","Viewer","Help","PPanel","Far","mf","Area","BM","eval","msgbox","prompt","Keys","exit","printf","mmode","akey","lshift","bnot","band","bor","rshift","bxor","far","win",
>     "IsLuaStateRecreated", --lua plugins
>     "_filename",           --scripts executed in luamacro via cmdline prefix
> 
>   };


порядок в танковых войсках! :)
